pkg/ccl/sqlproxyccl/forwarder.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Licensed as a CockroachDB Enterprise file under the Cockroach Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package sqlproxyccl
+
+import (
+	"context"
+	"net"
+
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/errors"
+)
+
+// ErrForwarderClosed indicates that the forwarder has been closed.
+var ErrForwarderClosed = errors.New("forwarder has been closed")
+
+// ErrForwarderStarted indicates that the forwarder has already started.
+var ErrForwarderStarted = errors.New("forwarder has already started")
+
+// forwarder is used to forward pgwire messages from the client to the server,
+// and vice-versa. At the moment, this does a direct proxying, and there is
+// no intercepting. Once https://github.com/cockroachdb/cockroach/issues/76000
+// has been addressed, we will start intercepting pgwire messages at their
+// boundaries here.
+type forwarder struct {
+	// ctx is a single context used to control all goroutines spawned by the
+	// forwarder.
+	ctx       context.Context
+	ctxCancel context.CancelFunc
+
+	// crdbConn is only set after the authentication phase for the initial
+	// connection. In the context of a connection migration, crdbConn is only
+	// replaced once the session has successfully been deserialized, and the
+	// old connection will be closed.
+	conn     net.Conn // client <-> proxy
+	crdbConn net.Conn // proxy <-> client
+
+	// errChan is a buffered channel that contains the first forwarder error.
+	// This channel may receive nil errors.
+	errChan chan error
+
+	mu struct {
+		syncutil.Mutex
+
+		// closed is set to true whenever the forwarder is closed explicitly
+		// through Close, or when any of its main goroutines has terminated,
+		// whichever that happens first.
+		//
+		// A new forwarder instance will have to be recreated if one wants to
+		// reuse the same pair of connections.
+		closed bool
+
+		// started is set to true once Run has been invoked on the forwarder.
+		// This is a safeguard to prevent callers from starting the forwarding
+		// process twice. This will never be set back to false.
+		started bool
```

Comment:
I like to avoid states that indicate partial construction like "started". Typically they can be avoided by redefining the constructor. E.x. instead of exposing newForwarder() and Run(), this could expose a Forward(...) (Forwarder, err) method that constructs the forwarder and contains Run's logic.

I think this would even allow us to get rid ef the err and panic in proxy_handler. Since Run only returns precodition errors and we can avoid them by making it part of the constructor.